action.go

```diff
@@ -280,6 +191,173 @@ func (c *cmdAction) String() string {
 	return "[" + strings.Join(ss, " ") + "]"
 }
 
+// CmdConfig allows creating Actions for commands using custom logic for
+// determining the ActionProperties without the need for creating a custom
+// Action implementation.
+//
+// This can be useful for working with custom commands provided by Redis
+// modules for which the built in logic may return sub-optimal properties or in
+// case some commands are known to be slow in some cases and therefore should
+// not use connection sharing.
+//
+// All methods on CmdConfig are safe for concurrent use from different
+// goroutines.
+type CmdConfig struct {
+	// ActionProperties is an optional callback that will be called when
+	// creating a new Action using CmdConfig.Cmd or CmdConfig.FlatCmd and is
+	// used to set the ActionProperties for the new Action.
+	//
+	// If ActionProperties is nil, a builtin default callback will be used that
+	// should work with all standard Redis commands but may not return correct
+	// results for custom commands provided by Redis Modules or unreleased
+	// commands.
+	ActionProperties func(cmd string, args ...string) ActionProperties
+}
+
+func defaultActionProperties(cmd string, args ...string) ActionProperties {
+	isBlocking := blockingCmds[strings.ToUpper(cmd)]
+	properties := ActionProperties{
+		CanRetry:     true,
+		CanPipeline:  !isBlocking,
+		CanShareConn: !isBlocking,
+	}
+	cmd = strings.ToUpper(cmd)
+	switch {
+	case noKeyCmds[cmd] || len(args) == 0:
+	case cmd == "BITOP" && len(args) > 1:
+		properties.Keys = args[1:]
+	case cmd == "XINFO":
+		if len(args) >= 2 {
+			properties.Keys = args[1:2]
+		}
+	case cmd == "XGROUP" && len(args) > 1:
+		properties.Keys = args[1:2]
+	case cmd == "XREAD" || cmd == "XREADGROUP":
+		properties.Keys = findStreamsKeys(args)
+	default:
+		properties.Keys = args[:1]
+	}
+	return properties
+}
+
+func findStreamsKeys(args []string) []string {
+	for i, arg := range args {
+		if strings.ToUpper(arg) != "STREAMS" {
+			continue
+		}
+
+		// after STREAMS only stream keys and IDs can be given and since there must be the same number of keys and ids
+		// we can just take half of remaining arguments as keys. If the number of IDs does not match the number of
+		// keys the command will fail later when send to Redis so no need for us to handle that case.
+		ids := len(args[i+1:]) / 2
+
+		return args[i+1 : len(args)-ids]
+	}
+
+	return nil
+}
+
+func (cfg CmdConfig) actionProperties(cmd string, args ...string) ActionProperties {
+	pf := cfg.ActionProperties
+	if pf == nil {
+		pf = defaultActionProperties
+	}
+	return pf(cmd, args...)
+}
+
+// Cmd works like the global Cmd function but using the value returned by
```

Comment:
I don't think it's necessary to re-document ActionProperties here, this could simple read:

```
Cmd works like the global Cmd function but can be additionally configured using
fields on CmdConfig. See the global Cmd's documentation for further details.
```

and similarly for FlatCmd

Reply:
Done